### PR TITLE
[Form] prevent duplicated error message for file upload limits

### LINKED ISSUE #32045: [Form] Double file size validation when file is beyond the 'upload_max_filesize' parameter
**Symfony version(s) affected**: 4.3.1

**Description**  
I have a double file size validation when file is beyond the 'upload_max_filesize' parameter.
In this case the max size is set to 500ko in the form and the max size allowed by php 2mo. Additionally the error message isn't translated.

sf-bug

**How to reproduce**  

    /**
     * Add or edit the avatar.
     */
    class AvatarFormType extends AbstractType
    {
        private const IMAGE_MAX_SIZE = 512000;  
        public function buildForm(FormBuilderInterface $builder, array $options): void
        {
            $builder->add('avatar', FileType::class, [
                'required' => true,
                'constraints' => [
                    new NotBlank(),
                    new Image([
                        'maxSize' => self::IMAGE_MAX_SIZE,
                    ]),
                ],
            ]);
        }
    }

and php setting : 
upload_max_filesize | 2M | 2M

**Possible Solution**  
Detect if a maxSize constraint already exists and is below the upload_max_filesize value?

See you. COil.

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #32045, #36503, #39107
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by yceruto — APPROVED** [review]

**Comment by thewalkingcoder** [comment]
@xabbuh thanks for your pull request, as expected i try with context #39107 same issue for me.

image

**Comment by xabbuh** [comment]
Thank you for the feedback. Could you create a small example application that allows me to reproduce and debug the issue?

**Comment by thewalkingcoder** [comment]
@xabbuh ok i try to create that today.

**Comment by thewalkingcoder** [comment]
@xabbuh 
you can try here ;-)
https://github.com/thewalkingcoder/uploadfile

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php [hunk 1]
```diff
@@ -124,6 +126,17 @@ public function mapViolation(ConstraintViolation $violation, FormInterface $form
 
         // Only add the error if the form is synchronized
         if ($this->acceptsErrors($scope)) {
+            if (File::TOO_LARGE_ERROR === $violation->getCode()) {
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php]: a better check could be `$violation->getCause() instanceof File` to account for the different error codes

**Comment by xabbuh** [comment]
Status: Needs work

**Comment by xabbuh** [comment]
@thewalkingcoder Thanks, based on your example I was able to find a few issue with my initial solution. Can you try the updated patch again?

File: src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php]: I don't think so. If there are multiple file upload files and one of them fails with another check of the `File` constraint we could potentially remove an error here that we would have to keep.

**Comment by xabbuh** [comment]
Status: Needs Review

**Comment by thewalkingcoder** [comment]
@xabbuh thanks for your works. 

Your commit resolve the duplicated message and so resolve #39107
image

 but if  i remove 
```
uploadIniSizeErrorMessage
```
```php
    /**
     * @var array<UploadedFile>

     * @Assert\NotBlank(message="Veuillez sélectionner un fichier.")
     * @Assert\All({
     *     @Assert\File(
     *         maxSize = "500k",
     *         maxSizeMessage = "Custom message max size.",
     *         uploadFormSizeErrorMessage = "Custom message max size form",
     *     )
     * })
     */
    public $files;
```
error message is

image

normally is

`
Custom message max size.
`

other issue ? or side effect with the PR ?

**Comment by xabbuh** [comment]
If I understand you correctly, that's expected. The file couldn't be uploaded as the maximum file limit configured for PHP is reached and thus the `FileValidator` will not perform any additional checks.

**Comment by thewalkingcoder** [comment]
@xabbuh i say just i don't understand the behaviour

| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert maxSize |  500ko |
| Assert  maxSizeMessage |  Custom Max file 500ko |
| Assert  uploadIniSizeErrorMessage | Custom message Max file php ini |
| file.pdf |  6mo |
| **result** | **Custom message Max file php ini** |

for me this PR resolve duplicated error Message.

but if i remove Assert uploadIniSizeErrorMessage 
| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert maxSize |  500ko |
| Assert  maxSizeMessage |  Custom Max file 500ko size |
| file.pdf |  6mo |
| **result** | the file is too large. Allowed maximum size is 0.5Mo |

default message maxSize is applied instead MaxSizeMessage


For me, the expected behaviour is

| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert maxSize |  500ko |
| file.pdf |  6mo |
| **result** | the file is too large. Allowed maximum size is 0.5Mo  (default message Assert MaxSize) |
| **behaviour** | ok actually |

| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert maxSize |  500ko |
| Assert  maxSizeMessage |  Custom Max file 500ko size |
| file.pdf |  6mo |
| **result** | Custom Max file 500ko size (custom message with maxSizeMessage) |
| **behaviour** | not the same actually |

| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert maxSize |  500ko |
| Assert  maxSizeMessage |  Custom Max file 500ko size |
| Assert  uploadIniSizeErrorMessage | Custom message Max file php ini |
| file.pdf |  6mo |
| **result** | Custom Max file 500ko size (custom message with maxSizeMessage)  |
| **behaviour** | not the same actually |

| config | value |
|-------|--------|
| php.ini |  2mo |
| Assert  uploadIniSizeErrorMessage | Custom message Max file php ini |
| Assert maxSize |  4mo |
| file.pdf |  6mo |
| **result** | Custom message Max file php ini (custom message with uploadIniSizeErrorMessage) |
| **behaviour** | ok actually |


I'm not sure is a side effect with this PR

**Comment by xabbuh** [comment]
Well, if the file exceeds the limit configured with `upload_max_filesize` the `uploadIniSizeErrorMessage` is used. If you do not configure it explicitly, its default value `The file is too large. Allowed maximum size is {{ limit }} {{ suffix }}.` will be used. The message configured with `maxSizeMessage` is only taken into account if the file size is below the limit configured with `upload_max_filesize`, but exceeds the `maxSize` value of the `File` constraint. So the results that you observe look expected to me.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by COil** [comment]
Hi, I can confirm it is fixed in 5.1.9. Thank you.

**Pull request merged**